include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,340 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ * 
+ */
+#ifndef INCLUDED_GHEX_UNSTRUCTURED_PATTERN_HPP
+#define INCLUDED_GHEX_UNSTRUCTURED_PATTERN_HPP
+
+#include <vector>
+#include <map>
+#include <numeric>
+#include <cstring>
+#include <iosfwd>
+
+#include "../transport_layer/context.hpp"
+#include "../allocator/unified_memory_allocator.hpp"
+#include "../pattern.hpp"
+#include "../buffer_info.hpp"
+#include "./grid.hpp"
+
+
+namespace gridtools {
+    namespace ghex {
+
+        /** @brief unstructured pattern specialization
+         *
+         * This class provides access to the receive and send iteration spaces, determined by the halos,
+         * and holds all connections to the neighbors.
+         *
+         * @tparam Communicator communicator type
+         * @tparam Index index type for domain and iteration space
+         * @tparam DomainId domain id type*/
+        template<typename Communicator, typename Index, typename DomainId>
+        class pattern<Communicator, unstructured::detail::grid<Index>, DomainId> {
+
+            public:
+
+                // member types
+                using communicator_type = Communicator;
+                using address_type = typename communicator_type::address_type;
+                using index_type = Index;
+                using grid_type = unstructured::detail::grid<index_type>;
+                using domain_id_type = DomainId;
+                using pattern_container_type = pattern_container<communicator_type, grid_type, domain_id_type>;
+
+                friend class pattern_container<communicator_type, grid_type, domain_id_type>;
+
+                /** @brief essentially a partition index and a sequence of remote indexes;
+                 * number of levels is provided as well, defaul is 1 (2D case):
+                 * the assumption is that each 2D element is a column, with 'levels' vertical elements.
+                 * WARN: a second index, and therefore a more complex iteration space,
+                 * is needed to handle correctly multiple vertical layers*/
+                class iteration_space {
+
+                    private:
+
+                        int m_partition;
+                        std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>> m_local_index;
+                        std::size_t m_levels;
+
+                    public:
+
+                        // ctors
+                        iteration_space() noexcept = default;
+                        iteration_space(const int partition,
+                                const std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>>& local_index,
+                                const std::size_t levels = 1) :
+                            m_partition{partition},
+                            m_local_index{local_index},
+                            m_levels{levels} {}
+                        iteration_space(const int partition,
```

Comment:
this constructor is not needed (duplicate of above - the const ref can bind to r-value refs)

Reply:
Yep, that's ok, I removed the second constructor. The only reason why it was there was to use the vector move semantics in case a r-value ref was passed to the constructor, but yes, this should never be the case, and anyway it should happen only in the setup phase.